### DISCUSSION TITLE
Hardhat artifacts, JSON and ABIs

### DISCUSSION THREAD

**Discussion by AlphaDio** [wevm/viem#1004] [General]
I have upgraded recently to 1.3.x, over 0.12.x, and I have to say I am quite perplexed by the way ABIs are handled by the library.

Hardhat produces artificats that contain ABIs and bytecodes under a JSON, I would think that viem would handle them directly, but the type inferrence you do ask that I make my ABI look a certain way, and that would involve taking the data from each of the artifacts I create with HardHat and put them in a file as const for viem to be able to handle them without sending out typescript errors.

Instead of being powerful and preventing errors, _all it does is make me waste copious amounts of time_, where with ethers.js, I could just import the json artificat, feed the Contract interface with the abi object and the bytecode object, and done.

Any possibilities to make the Contract Interfacing less strict? People like me use those frameworks to save time and build fast.

**Comment by jxom** [comment]
You can use the [wagmi CLI](https://wagmi.sh/cli/getting-started) with the [hardhat plugin](https://wagmi.sh/cli/plugins/hardhat) to generate ABIs cast to `as const` from your hardhat artifacts without copy + pasting.

Another note: you can still import ABIs as JSON like in wagmi 0.12 and ethers.js, however, the caveat is that you won’t get all the nice features of type safety & inference. In other words, casting to `as const` is only a feature addition (you don’t have to do it, though, it is highly recommended).

Hopefully importing JSON as const will be supported natively in TypeScript one day: https://github.com/microsoft/TypeScript/issues/32063

**Comment by mdesousa** [comment]
i was able to generate `abi.ts` by adding the subtask below to the hardhat config. it exports the abi as a constant, so it can be used as shown in [viem type inference](https://viem.sh/docs/typescript#type-inference). hope it helps!

```
type ContractMap = Record<string, { abi: object }>;

subtask(TASK_COMPILE_SOLIDITY_EMIT_ARTIFACTS).setAction(
  async (args, env, next) => {
    const output = await next();
    const { artifacts } = env.config.paths;
    const promises = Object.entries(args.output.contracts).map(
      async ([sourceName, contract]) => {
        const file = join(artifacts, sourceName, 'abi.ts');
        const { abi } = Object.values(contract as ContractMap)[0];
        const data = `export const abi = ${inspect(abi, false, null)} as const;`;
        await writeFile(file, data);
      },
    );
    await Promise.all(promises);
    return output;
  },
);
```

**Comment by CJ42** [comment]
You can generate the viem typed ABI in your Hardhat project using the following `wagmi.config.ts`

```ts
import { defineConfig } from '@wagmi/cli';
import fs from 'fs';

const artifacts = fs.readdirSync('./artifacts', {});

const contractsWagmiInputs = artifacts.map((artifact) => {
  const jsonArtifact = JSON.parse(fs.readFileSync(`./artifacts/${artifact}`, 'utf-8'));
  return {
    name: jsonArtifact.contractName,
    abi: jsonArtifact.abi,
  };
});

export default defineConfig({
  out: './abi.ts', // 👈🏻 configure output directory here
  contracts: contractsWagmiInputs,
  plugins: [],
});
```

You can then install `@wagmi/cli` and run the following command. It will output the typed ABIs to an `abi.ts` file at the root of your directory.

```bash
npm i --save-dev @wagmi/cli

npx wagmi generate
```
